presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayCombinationsFunction.java

```diff
@@ -0,0 +1,137 @@
+@ScalarFunction("combinations")
+public final class ArrayCombinationsFunction
+{
+    private ArrayCombinationsFunction() {}
+
+    private static final int MAX_COMBINATION_LENGTH = 5;
+    private static final int MAX_RESULT_ELEMENTS = 100_000;
+
+    @TypeParameter("T")
+    @SqlType("array(array(T))")
+    public static Block combinations(
+            @TypeParameter("T") Type elementType,
+            @SqlType("array(T)") Block array,
+            @SqlType(StandardTypes.INTEGER) long n)
+    {
+        int arrayLength = array.getPositionCount();
+        int combinationLength = toIntExact(n);
+        checkCondition(combinationLength >= 0, INVALID_FUNCTION_ARGUMENT, "combination size must not be negative: %s", combinationLength);
+        checkCondition(combinationLength <= MAX_COMBINATION_LENGTH, INVALID_FUNCTION_ARGUMENT, "combination size must not exceed %s: %s", MAX_COMBINATION_LENGTH, combinationLength);
+
+        ArrayType arrayType = new ArrayType(elementType);
+        if (combinationLength > arrayLength) {
+            return arrayType.createBlockBuilder(new PageBuilderStatus().createBlockBuilderStatus(), 0).build();
+        }
+
+        int combinationCount = combinationCount(arrayLength, combinationLength);
+        checkCondition(combinationCount * (long) combinationLength <= MAX_RESULT_ELEMENTS, INVALID_FUNCTION_ARGUMENT, "combinations exceed max size");
+
+        int[] ids = new int[combinationCount * combinationLength];
+        int idsPosition = 0;
+
+        int[] combination = firstCombination(combinationLength);
+        do {
+            arraycopy(combination, 0, ids, idsPosition, combinationLength);
+            idsPosition += combinationLength;
+        }
+        while (nextCombination(combination, arrayLength));
+        verify(idsPosition == ids.length, "idsPosition != ids.length, %s and %s respectively", idsPosition, ids.length);
+
+        int[] offsets = new int[combinationCount + 1];
+        setAll(offsets, i -> i * combinationLength);
+
+        return ArrayBlock.fromElementBlock(combinationCount, Optional.empty(), offsets, new DictionaryBlock(array, ids));
+    }
+
+    @VisibleForTesting
+    static int combinationCount(int arrayLength, int combinationLength)
+    {
+        try {
+            /*
+             * Then combinationCount(n, k) = combinationCount(n-1, k-1) * n/k (https://en.wikipedia.org/wiki/Combination#Number_of_k-combinations)
+             * The formula is recursive. Here, instead of starting with k=combinationCount, n=arrayLength and recursing,
+             * we start with k=0 n=(arrayLength-combinationLength) and proceed "bottom up".
+             */
+            int combinations = 1;
+            for (int i = 1; i <= combinationLength; i++) {
+                combinations = multiplyExact(combinations, arrayLength - combinationLength + i) / i;
+            }
+            return combinations;
+        }
+        catch (ArithmeticException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)